Fix memory errors

schoppmp opened the pull request:
This is an attempt to get ABY free of memory leaks. Currently, this includes only the main library, but not the example binaries. For the latter there needs to be a way to handle the share pointers that are returned by most circuit-building functions. One possible approach would be to wrap them in a new class that contains a C++11 smart pointer (i.e., a `std::shared_ptr<share>`). This would preserve compatibility with existing code and requires only minor changes to the examples. However, there might be other approaches that work better, so feel free to comment if you have an idea.

Right now, this PR depends on my fork of the OTExtension library. I'll change it back to the original repository as soon as the [corresponding PR](https://github.com/encryptogroup/OTExtension/pull/9) there is merged.

schoppmp committed:
Do not send uninitialized memory when deleting SndThread

schoppmp committed:
Delete receiver threads in ABYParty::Cleanup()

schoppmp committed:
Free OT extension memory

schoppmp committed:
Fix vector z size in AES S-Box

schoppmp committed:
Wait for threads instead of killing them

schoppmp committed:
Delete listener queues in ~RcvThread()

schoppmp committed:
Free more memory in ABYCircuit::Cleanup()

schoppmp committed:
Join worker threads when cleaning up ABYSetup

schoppmp committed:
Add destructor to YaoSharing

schoppmp committed:
Close more memory leaks in Yao{Server,Client}Sharing

schoppmp committed:
Close memory leaks in ABY{Party,Setup}

schoppmp committed:
Close memory leaks in SetupLUT

schoppmp committed:
Close more memory leaks

schoppmp committed:
Close memory leak in channel

schoppmp committed:
Only mark thread as not-running when it was joined

schoppmp committed:
Delete party channel after evaluating circuit in ABYParty

schoppmp committed:
Update OTExtension submodule

schoppmp committed:
Add boolean field indicating whether to clean up masking functions after running OT tasks

schoppmp committed:
Close memory leak in ABYPartyListen()

schoppmp committed:
Update OTExtension submodule

schoppmp committed:
Create a new masking object for each OT task

schoppmp committed:
Close memory leak in boolean conversion gates

lenerd commented:
Hey Phillipp,
thanks for your effort!

schoppmp committed:
Close memory leaks in boolean and Yao server sharing

dd23 commented:
Thanks a lot for the pull request! We're very grateful for your help with this ongoing issue.
We'll review your changes and will probably merge it within the next few days.

schoppmp committed:
Initialize m_vOutputDestionations in YaoServerSharing

schoppmp commented:
@dd23 
Don't worry too much about merging yet. As you can see, I'm still finding leaks and adding commits.
Also, I mostly concentrate on the parts that I intend to use myself, so I won't be looking into some of the more specialized features.
To find the remaining leaks, it would be good to have a comprehensive test suite, or at least an executable that uses all of the features that are currently implemented. As soon as there is a plan on how to handle dynamically allocated shares, we could fix the examples and maybe use them for testing, although I'm not sure how many of the features they cover. Do you think that makes sense?

schoppmp commented:
One more thing to keep in mind is platform dependance. I'm running Linux 4.11.5-1-ARCH with GCC 7.1.1 20170516 and Valgrind 3.12.0, but it would probably make sense to also test other distributions like Debian or Ubuntu.

schoppmp committed:
Mark combiner gates as used in boolean sharing

schoppmp committed:
Free all remaining gates in ABYCircuit::Reset()

dd23 commented:
@schoppmp 
Thanks for the heads up. I'll wait with the final merge. Maybe you can simply let me know when you think your changes have reached a somewhat stable state. OTOH you could just create a new PR, when you fix more stuff :)

**Re Tests:**
We have a testing executable in [https://github.com/encryptogroup/ABY/tree/public/src/test](https://github.com/encryptogroup/ABY/tree/public/src/test), that tests and verifies many of the internal functionalities and currently AES as one example. It it also checked by Travis CI (currently running Ubuntu 14.04) after every commit.
It should be relatively easy to extend this testing executable to include more features.

**Re Distros:**
Internally we use both Ubuntu and Manjaro (Arch-based) to develop, so those two Distros should be covered. We're also working on making ABY compatible with Windows, so this should be covered soon as well.

schoppmp commented:
@dd23 Thanks for pointing out the testing executable and Travis CI! Currently, there are still some errors when running the tests, but I am working on it.

The reason I opened this PR so early is to give you (and others) the chance to comment on some of the issues, especially on how to handle share pointers that are returned by the `Put*Gate` methods. In my own code, I currently use a wrapper class as described above, with implicit conversion and dereference operators:
```C++
/**
 * This class implements an adapter for ABY shares, essentially wrapping
 * a share pointer in a std::shared_ptr<share>.
 */

#pragma once
#include "abycore/circuit/share.h"
#include <memory>

class ABYShare {
public:
    // default constructor
    ABYShare() : inner(nullptr) {};
    // conversion constructor from share pointers
    ABYShare(share* inner) : inner(inner) {};
    // conversion operator to share pointers
    operator share*() {
        return inner.get();
    }
    // dereference operator
    share* operator->() {
        return inner.get();
    }
private:
    std::shared_ptr<share> inner;
};
```
Using that class, all that needs to change to make existing executables leak-free is to replace every `share *` variable with an `ABYShare`. Do you think it makes sense to include this class in ABY?
Another approach, for example, would be to refactor the `share` class to not use dynamic allocation in the first place, but that would probably break compatibility with existing code.

schoppmp committed:
Fix double-free in channel::blocking_receive

schoppmp committed:
Move UsedGate method do Sharing superclass

The memory that needs to be freed when calling UsedGate depends
only on the gate type, not the sharing from which the method is
called.

schoppmp committed:
Free masking function of SetupLUT OT tasks

schoppmp committed:
Fix truth table size in ABYCircuit::PutTruthTableGate

schoppmp committed:
Clean up remaining gates in ABYParty::Reset()

schoppmp commented:
I don't know why this test is failing... On my system, it passes all tests:
[runtest.txt](https://github.com/encryptogroup/ABY/files/1095474/runtest.txt)
So this is either some data race or the behaviour is platform-dependent. I'll try again tomorrow to see if I can get it to fail for me as well, but it would be good if someone else could double-check with a different machine.

dd23 commented:
I think there might be memory issues. I _assume_ these might come from the [OTExtension](https://github.com/encryptogroup/OTExtension) submodule.
There is a rudimentary [test binary](https://github.com/encryptogroup/OTExtension/blob/master/mains/test.cpp), that you can run. Mostly it works fine but sometimes, some test cases fail on my machine. 
I tried looking into it, but didn't have too much luck so far…

schoppmp committed:
Fix Windows version of thread.h

schoppmp commented:
I ran the tests of both ABY and OTExtension a couple of times in their currently published versions (i.e., commit [b83ec59](https://github.com/encryptogroup/OTExtension/commit/b83ec593f0134d2347f0f5e39ccc3b22610f50c7) for OTExtension and commit [9b4fc25](https://github.com/encryptogroup/ABY/commit/9b4fc258cee5293b9be4f7f09cd1168724563cc3) for ABY). For OTExtension, some of the tests fail, but not deterministically, so it's likely to be a race condition. I fixed a couple of issues in my fork, but some tests still fail from time to time. ABY also sometimes fails when running the PSI circuits, but I'm not sure whether that's a problem in ABY or in Phasing (after all, the latter has [stability problems](https://github.com/encryptogroup/ABY/issues/23) on its own).
Anyway, I think you're right, the main problem right now is the unreliability of the OTExtensions, so that needs to be addressed.

schoppmp committed:
Update OTExtension submodule

schoppmp committed:
Update OTExtension submodule

schoppmp committed:
Do not call Reset() in ABYPart::Cleanup()

schoppmp committed:
Fix out-of-bounds access in arithsharing

schoppmp commented:
@dd23 did you resolve the issue with the tests failing for this PR? Or why did you merge it? As I mentioned above, I'm not entirely sure why the tests here fail.

dd23 commented:
I could also not reproduce this error on my machine. I'm currently working on putting `util` to a separate repo that ABY and OTExt will include as submodule.
Since there are changes in `util`, I merged your changes to have a stable state for all repos.

However, I will investigate this issue on a different system later on.

dd23 commented:
I have pushed a [new branch](https://github.com/encryptogroup/ABY/tree/ENCRYPTO_utils) that uses the new [ENCRYPTO_utils repo](https://github.com/encryptogroup/ENCRYPTO_utils).

I'm currently having issues with the tests and will look into that issue very soon. Any help is appreciated, if someone wants to help.

dd23 commented:
Update: I've sorted out the biggest problems that resulted from separating the utils.
Some issues with threading and sockets remain and will be looked into very soon.
